Clean and document symex-dereference

smowton opened the pull request:
Review commit-by-commit. As you can see there were various bits of goto-symex where I pulled the threads to confirm what they did, and found the threads didn't go anywhere.

smowton committed:
Constify dereference_callbackt

has_failed_symbol is renamed to get_or_create_failed_symbol to emphasise that that method
really *may* have side-effects.

smowton committed:
Constify deref-state's back-pointer to goto_symex

This is only used to access its namespace in a const context.

smowton committed:
Goto symex: remove unused functions

smowton committed:
Remove unused dereference mode parameter

Both implementations currently disregard whether they're looking at an LHS or RHS expression.

smowton committed:
Symex-dereference: remove unused guard parameter

This was propagated between dereference_rec and address_arithmetic without ever changing,
before finally passing into value_set_dereferencet::dereference as a const reference.

smowton committed:
Document symex_dereference

tautschnig commented on src/pointer-analysis/value_set_dereference.h:
```diff
@@ -50,19 +50,14 @@ class value_set_dereferencet
   ///        be dereferenced.
   /// \param guard: A guard, which is assumed to hold when dereferencing.
   /// \param mode: Indicates whether the dereferencing is a load or store
   //    (unused).
```
Documentation needs to be cleaned up as well.

tautschnig commented on src/goto-symex/symex_dereference_state.cpp:
```diff
@@ -13,6 +13,21 @@ Author: Daniel Kroening, kroening@kroening.com
+/// \param expr: expression to get or create a failed symbol for
+/// \param [out] symbol: failed symbol result
+/// \return true if we populated \p symbol
 bool symex_dereference_statet::get_or_create_failed_symbol(
```
Not sure whether in this PR or in a separate one, but could we please change the return value to `optionalt<symbolt>`? The "Boolean return value + out-parameter" combination is just too error-prone and nowadays we have better options.

smowton replied:
Pushed an extra commit to just use `symbolt *`, since that's what the existing code was expecting

smowton committed:
Replace redundant pointer-and-boolean with just a pointer

It's already implicit in a bare pointer (rather than a reference) that it could be null,
so let's just use that to express whether we have an answer or not.

tautschnig commented:
This now needs a rebase (because I just merged the codet API PR) - sorry for the extra work!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 115cb7eaeda55e8c2269398e4a89eefb9e420e0f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100604500
